src/Npgsql/NpgsqlDataReader.cs

```diff
@@ -1741,25 +1740,61 @@ public override IEnumerator GetEnumerator()
         /// </summary>
         /// <returns></returns>
         public ReadOnlyCollection<NpgsqlDbColumn> GetColumnSchema()
-            => RowDescription == null || RowDescription.Fields.Count == 0
-                ? new List<NpgsqlDbColumn>().AsReadOnly()
-                : new DbColumnSchemaGenerator(_connection, RowDescription, _behavior.HasFlag(CommandBehavior.KeyInfo))
-                    .GetColumnSchema();
+            => GetColumnSchema(async: false).GetAwaiter().GetResult();
 
 #if !NET461
         ReadOnlyCollection<DbColumn> IDbColumnSchemaGenerator.GetColumnSchema()
             => new ReadOnlyCollection<DbColumn>(GetColumnSchema().Select(c => (DbColumn)c).ToList());
 #endif
 
+        /// <summary>
+        /// Asynchronously returns schema information for the columns in the current resultset.
+        /// </summary>
+        /// <returns></returns>
+#if NET // i.e. >= NET5_0
+        public new Task<ReadOnlyCollection<NpgsqlDbColumn>> GetColumnSchemaAsync(CancellationToken cancellationToken = default)
+#else
+        public Task<ReadOnlyCollection<NpgsqlDbColumn>> GetColumnSchemaAsync(CancellationToken cancellationToken = default)
+#endif
+        {
+            if (cancellationToken.IsCancellationRequested)
+                return Task.FromCanceled<ReadOnlyCollection<NpgsqlDbColumn>>(cancellationToken);
+            using (NoSynchronizationContextScope.Enter())
+                return GetColumnSchema(async: true, cancellationToken).AsTask();
+        }
+
+        ValueTask<ReadOnlyCollection<NpgsqlDbColumn>> GetColumnSchema(bool async, CancellationToken cancellationToken = default)
```

Comment:
As mentioned before, I'd make this a Task. This is a very low-perf API, 99% of usage will cause a Task instance to be allocated anyway (because the database needs to be contacted, and so the method completes asynchronously), and the single Task allocation is negligible.